docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+  requests/decisions (e.g. just as an eviction, an eviction with a hint about where to
+  reschedule, or as an eviction paired with a specific binding), how does the system
+  implement these requests, does the rescheduler take into account the second-order
+  effects of decisions (e.g. whether an evicted pod will reschedule, will cause
+  a preemption when it reschedules, etc.), does the rescheduler execute multi-step plans
+  (e.g. evict two pods at the same time with the intent of moving one into the space
+  vacated by the other, or even more complex plans)
+
+Additional musings on the rescheduling design space can be found [here](rescheduler.md).
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
+easy for an administrator to add new priority levels in between existing levels, to
+encourage thinking about priority in terms of user intent and avoid magic numbers, and to
+make the internal implementation more flexible.
```

Comment:
Yeah, @bgrant0607 had suggested some kind of user-specified priority relative to namespace. I think this might merit a separate sub-proposal, since it's a somewhat complicated topic. For example, I'm not sure how it interacts with quota, and also I'm not convinced you want to say some namespaces are always higher/lower-priority than some other namespaces. This is related to the question @mqliang asked later about whether preemption should be allowed at all across namespaces (my cop-out was that this should be a scheduler configuration pattern, but I thought most people would want the answer to be "yes").

BTW I didn't understand the "world" part in your comment. The way I was thinking about this stuff was analogous to: priority is a two-digit base-10 number, high-order digit is determined by namespace (mapping of namespace to digit is configured by cluster admin) and low-order digit is determined by user. Where does the "world" part come in?